oeps/oep-0026-arch-xapi-integration.rst

```diff
@@ -0,0 +1,483 @@
+from earlier experimental attempts on edX.org and elsewhere show how adaptive 
+learning mechanisms can significantly improve the learner's efficiency, engagement, 
+and retention, while providing a more effective learning experience with metacognitive
+support for life-long learning.
+
+.. _Adaptive Learning Presentation: https://openedx.atlassian.net/wiki/spaces/AC/pages/641105978/Adaptive+Learning+Presentation
+
+Integration with Adaptive Engines
+=================================
+
+As an iterative path to this future, the immediate goal is to enable integrations with
+external adaptive engines.  In the far future, we may also implement our own open-sourced
+adaptive engines.  But in the meantime, we would like to establish standard APIs
+that (1) adaptive engines can use to receive real-time events from the Open edX LMS and
+(2) the Open edX LMS can use to query adaptive engines on what to present to learners.
+This document focuses on #1 with the need for scalable and real-time updates to adaptive
+engines. This is depicted by the left-hand flow in the diagram below.
+
+.. image:: oep-0026/adaptive_learning_lms_basic.png
+   :alt: The left-hand side of the data flow diagram shows how the Open edX LMS sends
+    events to an intermediary "Eventing" component, which forwards those events to adaptive
+    engines. The right-hand side shows how the Open edX LMS queries the adaptive engines for
+    "what's next" via an intermediary "Adaptive Experiences" component.
+
+Integration with Other External Systems
+=======================================
+
+It is worthwhile to note that although the context and immediate requirements pertain to
+adaptive learning needs, the general framework for real-time notifications of Open edX LMS
+events can potentially be used by other external systems as well. However, for the time being,
+we will focus our efforts on the needs for adaptive learning, while keeping in mind that the
+technology and framework could potentially be extended and reused.
+
+
+Technical Specification
+-----------------------
+
+Standardized Integration with xAPI
+==================================
+In order to enable a plug-n-play ecosystem with various adaptive engines, we will
+use the ADL_ standard `Experience API (xAPI)`_ as the communication protocol between
+the Open edX LMS and registered adaptive engines. We have chosen xAPI since it has
+`widespread industry support and usage`_ for tracking and analyzing learning activity.
+Please see `xAPI Architecture Overview`_ for further information on xAPI.
+
+.. _ADL: https://adlnet.gov/
+.. _Experience API (xAPI): https://www.adlnet.gov/research/performance-tracking-analysis/experience-api/
+.. _widespread industry support and usage: https://xapi.com/adopters/
+.. _xAPI Architecture Overview: https://www.adlnet.gov/research/performance-tracking-analysis/experience-api/xapi-architecture-overview/
+
+Learning Record Store (LRS) - Future
```

Comment:
I think that the current development of xAPI needs to include (an) LRS. I think that when developing these structure and sends it to the engine it needs to keep somewhere (unless every engine will build his own LRS) and this data will be very valuable in the near future when more engines will arrive. (so they can train their model with your specific data sets). I don't think it is a lot of work to keep these sentences in a mongo collection for now (also we had some issues with the mongo collection - I hope to share more about that later) or to keep it in S3 (this is how we are implementing it today).

Reply:
Yes! We are currently testing learning locker (https://github.com/LearningLocker/learninglocker) and planning to use it for our developments. This will allow to further evaluate it and see if it can scale.

Reply:
Great - we used learning locker. The first version (in PHP) worked very slow and didn't support bulk insert of events. I think this version is better - did you try to have a load test on it?

Reply:
Not yet but you're right we should.

Reply:
Ok. I updated the LRS - Future section to indicate that we will not implement an LRS ourselves, but will look into integration with 3rd party implementations.